x/auction/keeper/auctions.go

```diff
@@ -0,0 +1,337 @@
+	if err != nil {
+		return a, err
+	}
+	for i, payout := range lotPayouts {
+		err = k.supplyKeeper.SendCoinsFromModuleToAccount(ctx, types.ModuleName, a.LotReturns.Addresses[i], sdk.NewCoins(payout))
+		if err != nil {
+			return a, err
+		}
+	}
+
+	// Update Auction
+	a.Bidder = bidder
+	a.Lot = lot
+	a.Bid = bid
+	// increment timeout
+	a.EndTime = earliestTime(ctx.BlockTime().Add(types.DefaultBidDuration), a.MaxEndTime)
+
+	return a, nil
+}
+func (k Keeper) PlaceBidReverse(ctx sdk.Context, a types.ReverseAuction, bidder sdk.AccAddress, lot sdk.Coin) (types.ReverseAuction, sdk.Error) {
+	// Validate New Bid
+	if lot.Denom != a.Lot.Denom {
+		return a, sdk.ErrInternal("lot denom doesn't match auction")
+	}
+	if lot.IsNegative() {
+		return a, sdk.ErrInternal("lot less than 0")
+	}
+	if !lot.IsLT(a.Lot) { // TODO add min bid decrements
+		return a, sdk.ErrInternal("lot not smaller than last lot")
+	}
+
+	// Move Coins
+	bidAmtToReturn := a.Bid
+	if bidder.Equals(a.Bidder) { // catch edge case of someone updating their bid with a low balance
+		bidAmtToReturn = sdk.NewInt64Coin(a.Bid.Denom, 0)
+	}
+	err := k.supplyKeeper.SendCoinsFromAccountToModule(ctx, bidder, types.ModuleName, sdk.NewCoins(bidAmtToReturn))
+	if err != nil {
+		return a, err
+	}
+	err = k.supplyKeeper.SendCoinsFromModuleToAccount(ctx, types.ModuleName, a.Bidder, sdk.NewCoins(bidAmtToReturn))
+	if err != nil {
+		return a, err
+	}
+
+	// Update Auction
+	a.Bidder = bidder
+	a.Lot = lot
+	// increment timeout
+	a.EndTime = earliestTime(ctx.BlockTime().Add(types.DefaultBidDuration), a.MaxEndTime)
+
+	return a, nil
+}
+
+// CloseAuction closes an auction and distributes funds to the highest bidder.
+func (k Keeper) CloseAuction(ctx sdk.Context, auctionID uint64) sdk.Error {
+
+	// get the auction from the store
+	auction, found := k.GetAuction(ctx, auctionID)
+	if !found {
+		return sdk.ErrInternal("auction doesn't exist")
+	}
+	// error if auction has not reached the end time
+	if ctx.BlockTime().Before(auction.GetEndTime()) {
+		return sdk.ErrInternal(fmt.Sprintf("auction can't be closed as curent block time (%v) is under auction end time (%v)", ctx.BlockTime(), auction.GetEndTime()))
+	}
+
+	// payout to the last bidder
+	var err sdk.Error
+	switch auc := auction.(type) {
+	case types.ForwardAuction, types.ForwardReverseAuction:
+		err = k.PayoutAuctionLot(ctx, auc)
+		if err != nil {
+			return err
+		}
+	case types.ReverseAuction:
+		err = k.MintAndPayoutAuctionLot(ctx, auc)
+		if err != nil {
+			return err
+		}
+	default:
+		panic("unrecognized auction type")
+	}
+
+	// Delete auction from store (and queue)
+	k.DeleteAuction(ctx, auctionID)
+	return nil
+}
+func (k Keeper) MintAndPayoutAuctionLot(ctx sdk.Context, a types.ReverseAuction) sdk.Error {
+	err := k.supplyKeeper.MintCoins(ctx, a.Initiator, sdk.NewCoins(a.Lot))
+	if err != nil {
+		return err
+	}
+	err = k.supplyKeeper.SendCoinsFromModuleToAccount(ctx, a.Initiator, a.Bidder, sdk.NewCoins(a.Lot))
```

Comment:
Should this be `a.GetBidder()`, `a.GetLot()`? It looks like auction's getter methods aren't fully leveraged in other functions as well.

Reply:
I switched things about to remove this. My intention was to only have getter and setters for code that deals with all auction types.

Reply:
Okay, that makes sense.